### DISCUSSION TITLE
[mini.statusline] Stop overriding the `laststatus` user option?

### DISCUSSION THREAD

**Discussion by ro0gr** [nvim-mini/mini.nvim#1120] [Ideas]
Right now, by default `mini.statusline` overrides the `laststatus` option and sets it to be `2`. This kinda of makes sense cause if you have a statusline plugin, you probably do not want the statusline being hidden.. However, I find this behavior confusing, cause it does silently overide an exlicit user configuration.

### Some background

I use the https://github.com/nvim-lua/kickstart.nvim, which includes the `mini.statusline` as a plugin enabled by default. I didn't even know that I use this plugin, until I realized my `laststatus` option set to `3` is not respected anymore. Note, it worked fine after re-sourcing my config but then it just **silently**_(and that's where I believe the problem is)_ stopped working after I reload my neovim instance. So for some time I didn't even realize I have my `laststatus` broken.

Since the kickstart.nvim does not have any preferences on the `laststatus`, I assume it's not the kickstart.nvim responsibility to make a decisiion about the [`set_vim_settings`](https://github.com/echasnovski/mini.nvim/blob/d8faae693d266ee4f3a2bf1fb7ca4bc81314c3e9/lua/mini/statusline.lua#L171-L174) option.

### What's the issue

My first attempt to address the issue was to respect the `laststatus=3` https://github.com/echasnovski/mini.statusline/pull/1. However, now I think the problem is really about mutating the `laststatus` at the first place. 

The current behavior makes configuring the `laststatus` in config pointless for the user. At the same time if I change the option via command line the value gets applied properly, so I can still hide the statusbar. This feels incosistent. 

I understand that there is a special explanation for the `laststatus=3` in the README. However, isn't it better if we don't need configuration for this at all?

### Suggestion

I'd like the plugin to not mutate the users global settings, and provide a user with an informative feedback in case of the "wrong" `laststatus` is detected on editor startup.

As far as I can tell the plugin works just fine with either `laststatus=2` and `laststatus=3`. So we could whitelist them and do nothing if one of them is recognized. Otherwise, we could just warn the user that the `laststatus` he has specified in the nvim config does not work nice with the plugin.

_Alternatively, instead of allowing "good" `laststatus` values we could maintain a list of "bad" values. However, it might cause issues with forward compatibility in case if new possible values would be added by the nvim in the future._

**Comment by ro0gr** [comment]
If the suggestion gets some support, I will work PR with a pleasure.

**Comment by echasnovski** [comment]
Thanks for the suggestion!

I already gave the answer in https://github.com/echasnovski/mini.statusline/pull/1:

> The documentation unambiguously and explicitly says what to do if user wants to use `laststatus = 3`. I think the current behavior is fine.

------

> The current behavior makes configuring the laststatus in config pointless for the user.

It does not. If user wants to use `laststatus = 3`, it is as straightforward as:

```lua
vim.o.laststatus = 3
require('mini.statusline').setup({ set_vim_settings = false })
```

> I'd like the plugin to not mutate the users global settings, and provide a user with an informative feedback in case of the "wrong" `laststatus` is detected on editor startup.

One of the main design goals of 'mini.nvim' modules is to make them work as much as reasonably possible after a single `require('mini.xxx').setup()`. Not ensuring proper `laststatus` goes against this.

------

At the moment, I don't think there is a need to change anything. But I'll take some time to maybe rethink.

**Reply by ro0gr** [reply]
Thanks for your reply!

Not arguing here, and I can definitely live with whatever is decided here, cause I found the reason of my confusion already. But I think this is not straightforward at all, and let me have another try to explain it from a slightly other angle:

Let's ignore my kickstart usecase(maybe not only mine) for a moment, and say I have the `mini.statusline`(and many others) installed for some time and working fine. Then, someday I decide to experiment with a plain neovim option for the `laststatus`. But suddenly it does not work. And now I have to figure out the plugin in the equation and do the special treatment from the README to disable the mutation behavior.

_Just trying to explain what's wrong IMO and why it isn't straighforward to have a dedicated option to restore the default neovim behavior._

**Comment by echasnovski** [comment]
Thanks again for the suggestion!

After thinking about it, making `set_vim_settings` be "ensure that statusline is always visible" is closer to what it was originally designed to do. So now on latest `main` branch its default `true` value results in setting `laststatus=2` only if it was 0 (never show) or 1 (show if there are more than one window).

**Reply by ro0gr** [reply]
Awesome! Thanks for the change 👍
